// @actions/exit/lib/exit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * The code to exit an action
 */
var ExitCode;
(function (ExitCode) {
    /**
     * A code indicating that the action was successful
     */
    ExitCode[ExitCode["Success"] = 0] = "Success";
    /**
     * A code indicating that the action was a failure
     */
    ExitCode[ExitCode["Failure"] = 1] = "Failure";
    /**
     * A code indicating that the action is complete, but neither succeeded nor failed
     */
    ExitCode[ExitCode["Neutral"] = 78] = "Neutral";
})(ExitCode = exports.ExitCode || (exports.ExitCode = {}));
// TODO: These exit codes may not behave as expected on the new runtime, due to
// complexities of async logging and sync exiting.
/**
 * Exit the action as a success.
 */
function success() {
    process.exit(ExitCode.Success);
}
exports.success = success;
/**
 * Exit the action as a failure.
 */
function failure() {
    process.exit(ExitCode.Failure);
}
exports.failure = failure;
/**
 * Exit the action neither a success or a failure
 */
function neutral() {
    process.exit(ExitCode.Neutral);
}
exports.neutral = neutral;
